File: iOS 8 by Tutorials/Objective-C/Chapter 19 - Beginning Photos/MultipleAssetPicker/MultipleImagePicker/AssetCollectionsViewController.m
//
//  RootTableViewController.m
//  Stitch
//
//  Created by Jack Wu on 2014-06-22.
//
//

@import Photos;

#import "AssetCollectionsViewController.h"
#import "AssetsViewController.h"

@interface AssetCollectionsViewController ()

@property (strong, nonatomic) NSArray * sectionNames;
@property (strong, nonatomic) PHFetchResult * userFavorites;
@property (strong, nonatomic) PHFetchResult * userAlbums;

@end

@implementation AssetCollectionsViewController

static NSString * const CollectionCellReuseIdentifier = @"CollectionCell";

#pragma mark - UIViewController

- (void)viewDidLoad {
  [super viewDidLoad];
    
  if (!self.selectedAssets) {
      self.selectedAssets = [[NSMutableArray alloc] initWithCapacity:6];
  }
  
  [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
    dispatch_async(dispatch_get_main_queue(), ^{
      if ([PHPhotoLibrary authorizationStatus] ==  PHAuthorizationStatusAuthorized) {
        [self fetchCollections];
        [self.tableView reloadData];
      }
      else {
        [self showNoAccessAlertAndCancel];
      }
    });
  }];

  self.sectionNames = @[@"", @"", @"Albums"];
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];
  [self.tableView reloadData];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  AssetsViewController * destination = (AssetsViewController *)segue.destinationViewController;
  
  destination.selectedAssets = self.selectedAssets;
  
  PHFetchOptions * options = [[PHFetchOptions alloc] init];
  options.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"creationDate" ascending:true]];
  NSIndexPath * indexPath = [self.tableView indexPathForCell:sender];
  if (indexPath.section == 0) {
    // Selected
    destination.assetsFetchResults = nil;
    destination.title = @"Selected";
  }
  else if (indexPath.section == 1) {
    if (indexPath.row == 0) {
      // All Photos
      destination.assetsFetchResults = [PHAsset fetchAssetsWithOptions:options];
      destination.title = @"All Photos";
    }
    else {
      // Favorites
      PHAssetCollection * favorites = self.userFavorites[indexPath.row - 1];
      destination.assetsFetchResults = [PHAsset fetchAssetsInAssetCollection:favorites options:options];
      destination.title = favorites.localizedTitle;
    }
  }
  else {
    // Albums
    PHAssetCollection * album = self.userAlbums[indexPath.row];
    destination.assetsFetchResults = [PHAsset fetchAssetsInAssetCollection:album options:options];
    destination.title = album.localizedTitle;
  }
}

#pragma mark - Private

- (void)fetchCollections {
  self.userFavorites = [PHAssetCollection fetchAssetCollectionsWithType:PHAssetCollectionTypeSmartAlbum subtype:PHAssetCollectionSubtypeSmartAlbumFavorites options:nil];
  self.userAlbums = [PHCollectionList fetchTopLevelUserCollectionsWithOptions:nil];
  
}

- (void)showNoAccessAlertAndCancel {
  UIAlertController * alert = [UIAlertController alertControllerWithTitle:@"No Photo Permissions" message:@"Please grant photo permissions in Settings"  preferredStyle:UIAlertControllerStyleAlert];
  [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
    [self cancelPressed:self];
  }]];
  [alert addAction:[UIAlertAction actionWithTitle:@"Settings" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
  }]];
  
}

#pragma mark - Actions

- (IBAction)donePressed:(id)sender {
  if ([self.delegate respondsToSelector:@selector(assetPickerDidFinishPickingAssets:)]) {
    [self.delegate assetPickerDidFinishPickingAssets:self.selectedAssets];
  }
}

- (IBAction)cancelPressed:(id)sender {
  if ([self.delegate respondsToSelector:@selector(assetPickerDidCancel)]) {
    [self.delegate assetPickerDidCancel];
  }
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return self.sectionNames.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  if (section == 0) {
    // Selected
    return 1;
  }
  else if (section == 1) {
    // All Photos + Favorites
    return 1 + self.userFavorites.count;
  }
  else {
    // User Albums
    return self.userAlbums.count;
  }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CollectionCellReuseIdentifier forIndexPath:indexPath];
  cell.detailTextLabel.text = @"";
  if (indexPath.section == 0) {
    // Selected
    cell.textLabel.text = @"Selected";
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)self.selectedAssets.count];
  }
  else if (indexPath.section == 1) {
    if (indexPath.row == 0) {
      // All Photos
      cell.textLabel.text = @"All Photos";
    }
    else {
      // Favorites
      PHAssetCollection * favorites = self.userFavorites[indexPath.row - 1];
      cell.textLabel.text = favorites.localizedTitle;
      if (favorites.estimatedAssetCount != NSNotFound) {
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)favorites.estimatedAssetCount];
      }
    }
  }
  else {
    // User Albums
    PHAssetCollection * album = self.userAlbums[indexPath.row];
    cell.textLabel.text = album.localizedTitle;
    if (album.estimatedAssetCount != NSNotFound) {
      cell.detailTextLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)album.estimatedAssetCount];
    }
  }
  return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
  return self.sectionNames[section];
}
@end

File: iOS 8 by Tutorials/Swift/Chapter 19 - Beginning Photos/MultipleAssetPicker-Starter/MultipleAssetPicker/AssetCollectionsViewController.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import UIKit
import Photos

protocol AssetPickerDelegate {
  func assetPickerDidFinishPickingAssets(selectedAssets: [PHAsset])
  func assetPickerDidCancel()
}

class AssetCollectionsViewController: UITableViewController {
  let AssetCollectionCellReuseIdentifier = "AssetCollectionCell"
  
  // MARK: Variables
  var delegate: AssetPickerDelegate?
  var selectedAssets: SelectedAssets?
  
  private let sectionNames = ["","","Albums"]
  private var userAlbums: PHFetchResult!
  private var userFavorites: PHFetchResult!
  
  // MARK: UIViewController
  override func viewDidLoad() {
    super.viewDidLoad()
    if selectedAssets == nil {
      selectedAssets = SelectedAssets()
    }
    
    // Check for permissions and load assets
  }
  
  override func viewWillAppear(animated: Bool)  {
    super.viewWillAppear(animated)
    tableView.reloadData()
  }
  
  override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject!) {
    let destination = segue.destinationViewController
      as AssetsViewController
    // Set up AssetCollectionViewController
  }
  
  // MARK: Private
  func fetchCollections() {

  }
  
  func showNoAccessAlertAndCancel() {
    let alert = UIAlertController(title: "No Photo Permissions", message: "Please grant Stitch photo access in Settings -> Privacy", preferredStyle: .Alert)
    alert.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: { action in
      self.cancelPressed(self)
    }))
    alert.addAction(UIAlertAction(title: "Settings", style: .Default, handler: { action in
      UIApplication.sharedApplication().openURL(NSURL(string: UIApplicationOpenSettingsURLString)!)
      return
    }))
    
    self.presentViewController(alert, animated: true, completion: nil)
  }
  
  // MARK: Actions
  @IBAction func donePressed(sender: AnyObject) {
    delegate?.assetPickerDidFinishPickingAssets(selectedAssets!.assets)
  }
  
  @IBAction func cancelPressed(sender: AnyObject) {
    delegate?.assetPickerDidCancel()
  }
  
  // MARK: UITableViewDataSource
  override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
    return sectionNames.count
  }
  
  override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    switch(section) {
    case 0: // Selected Section
      return 1
    case 1: // All Photos + Favorites
      return 1
    case 2: // Albums
      return 0
    default:
      return 0
    }
  }
  
  override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCellWithIdentifier(AssetCollectionCellReuseIdentifier, forIndexPath: indexPath) as UITableViewCell
    cell.detailTextLabel!.text = ""
    
    // Populate the table cell
    
    return cell
  }
}
